### PR TITLE
BugFix: TorchLayer now supports shot batching

### PR DESCRIPTION
**Pull request by astralcai**
**Context:**
`TorchLayer` raises an error when used with a shot vector.

**Description of the Change:**
- Adds support for shot batching to `TorchLayer`
- Fixes bug in `_to_qfunc_output_type` in `qnode.py` that turns the entire shot-batched output, as opposed to each output in the batch, into the qfunc output type.

**Related GitHub Issues:**
Fixes https://github.com/PennyLaneAI/pennylane/issues/5319
[sc-58099]

### CODE REVIEW

File: tests/gradients/finite_diff/test_finite_difference_shot_vec.py
Comment (astralcai) [inline tests/gradients/finite_diff/test_finite_difference_shot_vec.py:399]: Here it actually looks like the original tests were expecting incorrect output shapes. According to https://github.com/PennyLaneAI/pennylane/pull/5418, the level of priority for dimensions from outer dimension to inner dimension is:
```
1. Quantum Script in batch. No squeezing.
2. Shot choice in a shot vector. Squeezed if no shot vector.
3. Measurement in the quantum script. Squeezed out if only one measurement.
4. Parameter broadcasting. Squeezed out if no parameter-broadcasting.  Adds to array shape instead of adding tuple nesting.
5. Fundamental measurement shape.
```
Here for circuits `cost1` and `cost5`, the batch dimension 3 should be outside the dimension that corresponds to the number of measurements. For some reason, if the measurements are returned as a list, the dimension is not expected to be squeezed out even if the list contains only one item. Does anyone understand why?

Comment (astralcai) [reply tests/gradients/finite_diff/test_finite_difference_shot_vec.py:399]: Do you mean that I should create a separate issue and link it to this PR?

File: tests/qnn/test_qnn_torch.py
Comment (astralcai) [reply tests/qnn/test_qnn_torch.py:485]: Hmmmm, curious 🤔

File: tests/gradients/finite_diff/test_finite_difference_shot_vec.py
Comment (dwierichs) [reply tests/gradients/finite_diff/test_finite_difference_shot_vec.py:399]: As far as I understand, the code _actually_ produces what's called "expected" here, right?
So why does changing the "expected" not make the test fail. And if the "expected" as it currently is matches code behaviour but not the docs, we should fix that independently (but in particular we should fix it :sweat_smile: )

File: tests/qnn/test_qnn_torch.py
Comment (mudit2812) [reply tests/qnn/test_qnn_torch.py:485]: If you pass a fixture as a function argument then `usefixtures` isn't needed. `usefixtures` is more so for fixtures that don't return anything, such as fixtures that create a context, or fixtures for test setup/clean up.

File: tests/gradients/finite_diff/test_finite_difference_shot_vec.py
Comment (mudit2812) [reply tests/gradients/finite_diff/test_finite_difference_shot_vec.py:399]: Sounds like we need to discuss what we consider to be correct behaviour before any changes are merged.

> iirc, this is because we assume that if a user says [qml.probs([0, 1])] they really want a list-valued output for some reason.
If this is the case, then we need to update the documentation to reflect this.

**Review by mudit2812 — COMMENTED** [review]
Left a couple of initial comments. Seems to me that this is currently blocked by the expected shape of the returns.

File: tests/gradients/finite_diff/test_finite_difference_shot_vec.py
Comment (astralcai) [reply tests/gradients/finite_diff/test_finite_difference_shot_vec.py:399]: > As far as I understand, the code _actually_ produces what's called "expected" here, right? So why does changing the "expected" not make the test fail. And if the "expected" as it currently is matches code behaviour but not the docs, we should fix that independently (but in particular we should fix it 😅 )

Without my change, the code currently produces incorrect results, and the test is passing because it is expecting incorrect results. After my change, the tests failed, because now the code actually produces correct results, but the tests are expecting the wrong things. That's why I updated the tests.

Comment (astralcai) [reply tests/gradients/finite_diff/test_finite_difference_shot_vec.py:399]: Consider this circuit:
```
dev = qml.device("default.qubit", shots=[10, 10, 10, 10])

@qml.qnode(dev)
def circuit(inputs):
    qml.RX(inputs, wires=0)
    qml.RY(inputs, wires=1)
    qml.CNOT(wires=[0, 1])
    return [qml.expval(qml.PauliZ(0)), qml.expval(qml.PauliZ(1))]

x = np.zeros(3)
res = circuit(x)
```
If you examine the output shape, you see
```
>>> qml.math.shape(res)
(4, 2, 3)
```
Where the three dimensions correspond to shot vector length, number of measurements, parameter broadcasting, which follows the order of priority outlined in https://github.com/PennyLaneAI/pennylane/pull/5418. However, if you reduce the number of measurements to 1:
```
@qml.qnode(dev)
def circuit(inputs):
    qml.RX(inputs, wires=0)
    qml.RY(inputs, wires=1)
    qml.CNOT(wires=[0, 1])
    return [qml.expval(qml.PauliZ(0))]

x = np.zeros(3)
res = circuit(x)
```
This happens:
```
>>> qml.math.shape(res)
(1, 4, 3)
```
See that the dimension for the number of measurements is placed outside the dimension of the shot vector, which violates the return type specifications, and also inconsistent.

File: pennylane/qnn/torch.py
Comment (astralcai) [reply pennylane/qnn/torch.py:439]: If there is shot batching, `res` would be a tuple, and `_combine_dimensions` is called on each result, so the dimension for shot batching would be the outermost dimension here.

**Comment by astralcai** [comment]
> what is going on in finite diff and SPSA with the permuted axes
@dwierichs Can you elaborate?

File: pennylane/qnn/torch.py
Comment (dwierichs) [reply pennylane/qnn/torch.py:439]: Thanks!

**Comment by dwierichs** [comment]
@astralcai  I was just referring to the output shape question for these gradient methods. 
I do agree that there seems to be a discrepancy between the output shapes produced by PennyLane and the technical specification doc. But I am a bit surprised that this only surfaces in two specific gradient method tests and it is a bug fix unrelated to the original PR. Might be nice to split this off and e.g. add forward pass tests that detect this difference.

**Comment by astralcai** [comment]
> @astralcai I was just referring to the output shape question for these gradient methods. I do agree that there seems to be a discrepancy between the output shapes produced by PennyLane and the technical specification doc. But I am a bit surprised that this only surfaces in two specific gradient method tests and it is a bug fix unrelated to the original PR. Might be nice to split this off and e.g. add forward pass tests that detect this difference.

This only happens when the measurements are in a list. If you change the list to a tuple, the error does not appear, which is the case for most tests out there.

**Comment by dwierichs** [comment]
Ahh, I understand now, thank you! :pray: Sorry, I overlooked that distinction.

 I still think we should test a forward-pass output shape not only within differentiation tests :) Would you mind adding a tiny test for this? If the code is ever changed, it would be confusing not to have any forward pass tests fail, but two differentation tests.

File: doc/releases/changelog-dev.md
Comment (dwierichs) [inline doc/releases/changelog-dev.md:479]: Would you mind also adding the return type fix? :see_no_evil: Sorry for being picky about this, I think it's good to have this change tracked :)

**Review by dwierichs — APPROVED** [review]
Thanks @astralcai :tada: 
Just small suggestions to wrap up, but looks good to me!

File: tests/qnn/test_qnn_torch.py
Comment (astralcai) [reply tests/qnn/test_qnn_torch.py]: Done

File: tests/test_return_types_qnode.py
Comment (astralcai) [reply tests/test_return_types_qnode.py]: Done

**Pull request merged**
